src/DI/ContainerBuilder.php

```diff
@@ -59,14 +62,23 @@ class ContainerBuilder
 	 */
 	public function addDefinition($name, ServiceDefinition $definition = NULL)
 	{
+		$this->definitionsChanged = TRUE;
 		if (!is_string($name) || !$name) { // builder is not ready for falsy names such as '0'
 			throw new Nette\InvalidArgumentException(sprintf('Service name must be a non-empty string, %s given.', gettype($name)));
 		}
 		$name = isset($this->aliases[$name]) ? $this->aliases[$name] : $name;
 		if (isset($this->definitions[$name])) {
 			throw new Nette\InvalidStateException("Service '$name' has already been added.");
 		}
-		return $this->definitions[$name] = $definition ?: new ServiceDefinition;
+		if (!$definition) {
+			$definition = new ServiceDefinition();
+		}
+		$definition->onChange[] = function ($service, $method) {
+			if (in_array($method, ['setClass', 'setFactory', 'setImplement', 'setAutowired'], TRUE)) {
+				$this->definitionsChanged = TRUE;
```

Comment:
Definitions changed even with the other methods. Maybe rename it to sth like classListObsolete?